rpm/files.go

```diff
@@ -0,0 +1,118 @@
+package rpm
+
+import (
+	"context"
+	"fmt"
+	"io/fs"
+	"sync"
+
+	"github.com/quay/claircore"
+	"github.com/quay/zlog"
+)
+
+// filesCache is used for concurrent access to the map containing
+// [claircore.Layer] -> map[string]struct{}.
+type filesCache struct {
+	c    map[string]map[string]struct{}
+	refs map[string]int
+	mu   sync.Mutex
+}
+
+var fc = &filesCache{
+	c:    map[string]map[string]struct{}{},
+	refs: map[string]int{},
+}
+
+// gc decrements the reference counter and deletes the [claircore.Layer]'s
+// entry from the cache map and the reference counter once the references
+// are 0 (or less).
+func (fc *filesCache) gc(key string) {
+	fc.mu.Lock()
+	defer fc.mu.Unlock()
+	fc.refs[key]--
+	if fc.refs[key] <= 0 {
+		delete(fc.c, key)
+		delete(fc.refs, key)
+	}
+}
+
+// get increments the reference counter for the key and looks up the key
+// in the cache. It will subsequently start a goroutine to track when the
+// calling context is done, calling the gc method.
+func (fc *filesCache) get(ctx context.Context, key string) (map[string]struct{}, bool) {
+	fc.mu.Lock()
+	defer fc.mu.Unlock()
+	fc.refs[key]++
+	f, ok := fc.c[key]
+	go func() {
+		<-ctx.Done()
+		fc.gc(key)
+	}()
+	return f, ok
+}
+
+// set sets the files for a particular key and deals with the locking.
+func (fc *filesCache) set(key string, files map[string]struct{}) {
+	fc.mu.Lock()
+	defer fc.mu.Unlock()
+	fc.c[key] = files
+}
+
+// getFiles looks up RPM files that exist in the RPM database using the
+// filesFromDB function and memorizes the result to avoid repeated work
+// for the same [claircore.Layer].
+func (fc *filesCache) getFiles(ctx context.Context, layer *claircore.Layer) (map[string]struct{}, error) {
+	if fc == nil {
+		panic("programmer error: filesCache nil")
+	}
+	key := layer.Hash.String()
+	if files, ok := fc.get(ctx, key); ok {
```

Comment:
I'm concerned about the scenario where we call `fc.get(ctx, key)` a bunch of times before we ever get to call `fc.set(key, files)`, which would nullify the cache.

I think we can throw lines [73, 106] into a function and use it for a `(*singleflight.Group).DoChan`. Something like:

```
ch := fc.sf.DoChan(key, findFiles(ctx, key, layer))
select {
case <-ctx.Done():
	return nil, ctx.Err()
case res := <-ch:
	return res.Val.(map[string]struct{}), res.Err
}
```

This way, we only ever get the layer's files once until its entry is deleted from the cache

Reply:
Thanks for the suggestion, I re-worked this a little, I'm kind of between `Do` and `DoChan` I guess the context cancelling before the result is returned is useful